src/gif.rs

```diff
@@ -81,19 +81,28 @@ impl<R> Read for GifReader<R> {
 impl<'a, R: 'a + Read> ImageDecoder<'a> for Decoder<R> {
     type Reader = GifReader<R>;
 
-    fn dimensions(&self) -> (u64, u64) {
-        (u64::from(self.reader.width()), u64::from(self.reader.height()))
+    fn dimensions(&self) -> (u32, u32) {
+        (u32::from(self.reader.width()), u32::from(self.reader.height()))
     }
 
     fn color_type(&self) -> color::ColorType {
         color::ColorType::Rgba8
     }
 
     fn into_reader(self) -> ImageResult<Self::Reader> {
-        Ok(GifReader(Cursor::new(self.read_image()?), PhantomData))
+        if self.total_bytes() > usize::max_value() as u64 {
+            return Err(ImageError::InsufficientMemory);
+        }
+
+        let mut buf = vec![0; self.total_bytes() as usize];
+        self.read_image(&mut buf)?;
```

Comment:
This seems like a duplicate of the above. Maybe it should be a utility or a default implemented extension method.